---
title: "A fast {\tt R} program to detect allelic imbalances using SNP array data with MAD"
subtitle: "Juan R. Gonzalez"
author: |
  | Institute for Global Health (ISGlobal), Barcelona, Spain
  | Bioinformatics Research Group in Epidemiolgy (BRGE)
  | (<http://brge.isglobal.org>)
date: "`r Sys.Date()`"
package: "`r pkg_ver('mad')`"
output: 
  BiocStyle::pdf_document:
    number_sections: true
    toc: yes
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Mosaic alteation detection with MAD}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```


# Introduction
MAD (mosaic alteration detector) is a fast algorithm to detect allelic imbalances using SNP array data (B-allele frequency (BAF) and log R ratio (LRR). The method detects different types of mosaic alterations (UPDs, deletions, duplications and trisomies) by searching for segments were B-deviation is different from 0. Segmentation procedure is carried out using an _R_ implementation of GADA algorithm (Pique-Regi R, Caceres A, Gonzalez JR. BMC Bioinformatics, 2010). R-GADA is available at \url{https://github.com/isglobal-brge/R-GADA}. 

Herein, we briefly describe how to use the main functions by using a real data set of 5 samples ( _CASE369.txt_,  _CASE371.txt_,  _CASE377.txt_,  _CONTROL152.txt_,  _CONTROL191.txt_) that can be downloaded from \url{http://www.creal.cat/media/upload/arxius/jr/GADA/MAD.zip}.


# Getting Started}
The package depends on R-GADA package that can be installed from our GitHub repository:

```{r install_gada, eval=FALSE}
devtools::install_github("isglobal-brge/R-GADA")
```

After that, MAD can be installed by

```{r install_mad, eval=FALSE}
devtools::install_github("isglobal-brge/MAD")
```


Data is loaded as usually

```{r load_mad}
library(mad)
``` 

# Required data format
The package enforces a strict directory structure on the working directory to perform the analysis of multiple samples. Nonetheless, the only required directory to be set up by the user is that containing the raw data. This is an example of how data are organized before the analysis is completed

|-- rawData
|   |-- CASE369.txt 
|   |-- CASE371.txt
|   |-- CASE377.txt
|   |-- CONTROL152.txt
|   |-- CONTROL191.txt


Notice that _rawData_ directory contains all data files corresponding to each individual from a particular assay. Each text file must be prepared in pennCNV format \url{http://http://penncnv.openbioinformatics.org}. This can be performed by using the BeadStudio tool when analyzing Illumina data (see \url{http://www.illumina.com/}), or other Bioconductor packages if Affymetrix SNP data is analyzed ( _crlmm_, _affy2sv_, ...). This is the structure of a given  example: 

  
```
Name        Chr Position  Log.R.Ratio B.Allele.Freq   GType
rs758676	  7	  12878632	0.1401	    0.4977	        AB
rs3916934	  13	103143536	0.3934	    0.4610	        AA
rs2711935	  4	  38838852	-0.1091	    0.0026	        AA
rs17126880	1	  64922104	0.0478	    0.9910	        AA
rs12831433	12	4995220	  -0.1661	    0.0000	        AA
```

\textcolor{red} {It is important to mention} that all files included in the folder _rawData_ must belong to the same type of array (e.g. same number of probes) because the annotation data is obtained from one of these files. If samples have been analyzed using different platforms, the best option is to create another working directory containing another _rawData_ folder. 

# Importing a collection of array data in pennCNV format

Let us prepare a `rawData` folder containing 5 files in the required format. PennCNV files are available in our package called `brgedata` that can be installed by executing:

```{r install_brgedata, eval=FALSE}
devtools::install_github("isglobal-brge/brgedata")
```

Once `brgedata` package is available in your computer we can copy the required files into a `rawData` folder by executing

```{r copy_data}
ss1 <- system.file("extdata/madData", package="brgedata")
dir.create("rawData")
ss2 <- "rawData"
files <- list.files(ss1)
file.copy(file.path(ss1,files), ss2)
```

As a result, our working directory contains a folder called `rawData` containing all the files we want to process. 

```{r show_wd}
dir()
dir("rawData")
``` 


Therefore, we are ready to start the analyses. First, data can be prepared for the analysis by:
 
```{r setup_mad}
example <- setupParGADA.B.deviation(NumCols=6, GenoCol=6, 
                                    BAFcol=5, log2ratioCol=4)
``` 

At it can be seen, the order of the columns in the pennCNV files is only mandatory for the three fist columns. The Genotypes, BAF and LRR can be changed by using _GenoCol_, _BAFcol_ and _log2ratioCol_. The function also contains an argument called _folder_ that allow the user to indicate where the _rawData_ is located

The object _example_ contains the following information:

```{r show_setup_mad}
example
```

#  Segmentation procedure

Segmentation is performed by using two consecutive algorithms implemented in two separate R-functions within the `gada` package. The first function (`parSBL`) uses sparse Bayesian learning (SBL) to discover the most likely positions and magnitudes for a segment, i.e. the breakpoints. The SBL model is governed by a hierarchical Bayesian prior, which is uninformative with respect to the location and magnitude of the copy number changes but restricts the total number of breakpoints. Sensitivity, given by the maximum breakpoint
sparseness, is controlled by the hyperparameter `aAlpha`. The second function (`parBE`) is an algorithm that uses a backward elimination (BE) strategy to rank the statistical significance of each breakpoint obtained from `SBL`. We have programmed another function called `parBE.B.deviation` to detect segments that can be considered as an allelic imbalance. The results from `parSBL` and `parBE.B.deviation` are stored in separate files, one for each sample, in a folder called `SBL`. 

Therefore, the segmentation procedure can be performed by executing next two steps:
  
```{r parSBL}
parSBL(example, estim.sigma2=TRUE, aAlpha=0.8)
```

and

```{r parBdev}
parBE.B.deviation(example, T=9, MinSegLen=75)
``` 

As previously mentioned, the parameter `aAlpha` controls the number of breakpoints. We consider that `aAlpha=0.8` is a good choice for Illumina 1M. The parameter `T` controls the False Discovery Rate (FDR). We are currently working on estimating the FDR but the recommended settings of `aAlpha` and `T` depending on the desired level of sensitivity and FDR for a Illumina 650K data are.

\begin{table}[h]
\begin{center}
\begin{tabular}{r c c }
\hline
(higher sensitivity , higher FDR ) & $<-->$ & ($a_{\alpha}=0.2$,$T>3$)\\
& $<-->$ & ($a_{\alpha}=0.5$,$T>4$)\\
(lower sensitivity ,  lower FDR ) & $<-->$ & ($a_{\alpha}=0.8$,$T>5$)\\
\hline
\end{tabular}
\end{center}
\end{table}


Anyway the backward elimination procedure is very fast and the user can change the parameter `T` and obtain different results. The argument `MinSegLen` indicates the number of consecutive probes that have a B-deviation different from 0.

# Obtaining allelic imbalance regions

This function creates a file with the following information:
  
```{r export2file}
exportSegments2File(example, file="example_T_9.txt")
``` 

The column `State` is a preliminary attempt to classify mosaic alterations after MAD calling process based on log2-ratio segment values (LRR) together with the percentage of normal heterozygous (BAF $\sim$ 0.5) and homozygous probes. The number codes correspond to the following abnormalities: UPD (1), deletion (2), duplication (3), trisomy (4) and LOH (5). It is recommendable to check called segments by making chromosome plots and observing LRR and BAF values to confirm those segments with limit cut-off values.  


Other regions are found if one changes the `T` parameter


```
> parBE.B.deviation(example, T=7, MinSegLen=75) 
```

and then execute

```
> exportSegments2File(example, file="example_T_7.txt") 
```
 
Finally, if the user is interested in looking a given region with regard to B-allele frequency and `log2ratio`, the  `plotChr` function can be used to get the figure \ref{fig-control191chr20}


```{r control191chr20}
> plotChr(example, sample="CONTROL191", chr=20)
```


```{r remove directories, echo=FALSE}
unlink("rawData", recursive = TRUE)
unlink("SBL", recursive = TRUE)
```

```{r sessionInfo}
sessionInfo()
```