---
title: "A fast {\tt R} program to detect allelic imbalances using SNP array data with MAD"
subtitle: "Juan R. Gonzalez"
author: |
  | Institute for Global Health (ISGlobal), Barcelona, Spain
  | Bioinformatics Research Group in Epidemiolgy (BRGE)
  | (<http://brge.isglobal.org>)
date: "`r Sys.Date()`"
package: "`r pkg_ver('mad')`"
output: 
  BiocStyle::pdf_document:
    number_sections: true
    toc: yes
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Mosaic alteation detection with MAD}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```


# Introduction
MAD (mosaic alteration detector) is a fast algorithm to detect allelic imbalances using SNP array data (B-allele frequency (BAF) and log R ratio (LRR). The method detects different types of mosaic alterations (UPDs, deletions, duplications and trisomies) by searching for segments were B-deviation is different from 0. Segmentation procedure is carried out using an _R_ implementation of GADA algorithm (Pique-Regi R, Caceres A, Gonzalez JR. BMC Bioinformatics, 2010). R-GADA is available at \url{https://github.com/isglobal-brge/R-GADA}. 

Herein, we briefly describe how to use the main functions by using a real data set of 5 samples ( _CASE369.txt_,  _CASE371.txt_,  _CASE377.txt_,  _CONTROL152.txt_,  _CONTROL191.txt_) that can be downloaded from \url{http://www.creal.cat/media/upload/arxius/jr/GADA/MAD.zip}.


# Getting Started}
The package depends on R-GADA package that can be installed from our GitHub repository:

```{r install_gada, eval=FALSE}
devtools::install_github("isglobal-brge/R-GADA")
```

After that, MAD can be installed by

```{r install_mad, eval=FALSE}
devtools::install_github("isglobal-brge/MAD")
```


Data is loaded as usually

```{r load_mad}
library(mad)
``` 

# Required data format
The package enforces a strict directory structure on the working directory to perform the analysis of multiple samples. Nonetheless, the only required directory to be set up by the user is that containing the raw data. This is an example of how data are organized before the analysis is completed

|-- rawData
|   |-- CASE369.txt 
|   |-- CASE371.txt
|   |-- CASE377.txt
|   |-- CONTROL152.txt
|   |-- CONTROL191.txt


Notice that _rawData_ directory contains all data files corresponding to each individual from a particular assay. Each text file must be prepared in pennCNV format \url{http://http://penncnv.openbioinformatics.org}. This can be performed by using the BeadStudio tool when analyzing Illumina data (see \url{http://www.illumina.com/}), or other Bioconductor packages if Affymetrix SNP data is analyzed ( _crlmm_, _affy2sv_, ...). This is the structure of a given  example: 

  
```{r show_example}  
read.delim("rawData/CASE369.txt", nrows = 5)
```

\textcolor{red} {It is important to mention} that all files included in the folder _rawData_ must belong to the same type of array (e.g. same number of probes) because the annotation data is obtained from one of these files. If samples are going to be analyzed using different platforms, the best option is to create another working directory containing another _rawData_ folder. 

# Importing a collection of array data in pennCNV format

If the working directory contains the _rawData_ folder, all samples can be prepared for the analysis by:
 
```{r setup_mad}
ff <- 
example <- setupParGADA.B.deviation(NumCols=6, GenoCol=6, 
                                    BAFcol=5, log2ratioCol=4)
``` 

At it can be seen, the order of the columns in the pennCNV files is only mandatory for the three fist columns. The Genotypes, BAF and LRR can be changed by using _GenoCol_, _BAFcol_ and _log2ratioCol_. The function also contains an argument called _folder_ that allow the user to indicate where the _rawData_ is located

The object _example_ contains the following information:

```{r show_setup_mad}
example
```
